File: cp_rust/src/bin/cses-1070-permutations.rs
// Created by Ayush Biswas at 2025/07/09 22:34
// https://cses.fi/problemset/task/1070
use cp_lib::*;

// @code begin
use cpio::*;
use std::iter::once;

sol! {
    fn solution(
        n: usize,
    ) -> CPResult<Words<usize>, String> {
        if n == 1 {
            return Success(once(1).collect())
        }
        if n <= 3 {
            return Failure("NO SOLUTION".into())
        }

        Success((2..=n).step_by(2).chain((1..=n).step_by(2)).collect())
    }
}

// @code end

File: cp_haskell/bin/cses-1070-permutations.hs
-- Created by Ayush Biswas at 2026/03/03 15:54
-- https://cses.fi/problemset/task/1070

-- @code begin
import Text.Printf (printf)

sol :: Int -> [Int]
sol n = [2, 4 .. n] ++ [1, 3 .. n]

main :: IO ()
main = do
  n <- read <$> getLine
  solution n
  where
    solution n
      | n == 1 = print 1
      | n < 4 = putStrLn "NO SOLUTION"
      | otherwise = mapM_ (printf "%d ") $ sol n

-- @code end